File: cpp/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/SurfaceRayIntersection.cpp
#include "vsdk/toolkit/common/linealAlgebra/Vector3Dd.h"
#include "vsdk/toolkit/environment/geometry/Geometry.h"
#include "vsdk/toolkit/environment/geometry/element/Ray.h"
#include "vsdk/toolkit/environment/geometry/element/RayHit.h"
#include "vsdk/toolkit/environment/geometry/surface/FunctionalExplicitSurface.h"
#include "vsdk/toolkit/environment/geometry/surface/TriangleMesh.h"
#include "vsdk/toolkit/environment/geometry/surface/TriangleMeshGroup.h"
#include "vsdk/toolkit/environment/geometry/volume/Box.h"
#include "vsdk/toolkit/environment/geometry/geometricProcessing/SurfaceRayIntersection.h"
static bool rayIntersectsGeometryBounds(Geometry* geometry, const Ray& inRay)
{
    double* mm = geometry->getMinMax();
    Vector3Dd size(mm[3]-mm[0], mm[4]-mm[1], mm[5]-mm[2]);
    Vector3Dd center(
        (mm[3]+mm[0])/2,
        (mm[4]+mm[1])/2,
        (mm[5]+mm[2])/2
    );
    delete [] mm;

    Box boundingVolume(size);
    Ray localRay(
        inRay.getOrigin().subtract(center),
        inRay.getDirection(),
        inRay.getT()
    );
    return boundingVolume.doIntersection(localRay, 0);
}

bool SurfaceRayIntersection::doIntersection(Geometry* geometry, const Ray& inRay, RayHit* outHit)
{
    if ( geometry == 0 ) {
        return false;
    }

    if ( dynamic_cast<TriangleMesh*>(geometry) != 0 ||
         dynamic_cast<TriangleMeshGroup*>(geometry) != 0 ||
         dynamic_cast<FunctionalExplicitSurface*>(geometry) != 0 ) {
        if ( !rayIntersectsGeometryBounds(geometry, inRay) ) {
            return false;
        }
    }

    return geometry->doIntersection(inRay, outHit);
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/SurfaceRayIntersection.ts
import { Vector3Dd } from "../../../common/linealAlgebra/Vector3Dd.js";
import { Geometry } from "../Geometry.js";
import { Ray } from "../element/Ray.js";
import { RayHit } from "../element/RayHit.js";
import { FunctionalExplicitSurface } from "../surface/FunctionalExplicitSurface.js";
import { TriangleMesh } from "../surface/TriangleMesh.js";
import { TriangleMeshGroup } from "../surface/TriangleMeshGroup.js";
import { Box } from "../volume/Box.js";

export class SurfaceRayIntersection {
    public static doIntersectionFirstHit(geometry: Geometry | null, ray: Ray, out: RayHit): boolean {
        if (geometry === null) return false;
        if (
            (geometry instanceof TriangleMesh ||
                geometry instanceof TriangleMeshGroup ||
                geometry instanceof FunctionalExplicitSurface) &&
            !this.rayIntersectsGeometryBounds(geometry, ray)
        )
            return false;
        return geometry.doIntersectionFirstHit(ray, out);
    }
    private static rayIntersectsGeometryBounds(geometry: Geometry, ray: Ray): boolean {
        const mm = geometry.getMinMax();
        const size = new Vector3Dd(mm[3]! - mm[0]!, mm[4]! - mm[1]!, mm[5]! - mm[2]!);
        const center = new Vector3Dd((mm[3]! + mm[0]!) / 2, (mm[4]! + mm[1]!) / 2, (mm[5]! + mm[2]!) / 2);
        return new Box(size).doIntersectionFirstHit(
            new Ray(ray.getOrigin().subtract(center), ray.getDirection(), ray.getT()),
            new RayHit(),
        );
    }
}
